bpo-38483: Add support for venv.ini

cooperlees opened the pull request:
- Add argparse.ArgumentDefaultsHelpFormatter so user can see overloads
- Add new CliDefaults NamedTuple with defaults
- Check for ~/.venvrc + parse if exists
- Check args, if true or false bool otherwise string
- Overload by creating NamedTuple passing kwargs
- Update docs

Test:
- Run `./python.exe -m venv --help`
- Add unit tests for new function


<!-- issue-number: [bpo-38483](https://bugs.python.org/issue38483) -->
https://bugs.python.org/issue38483
<!-- /issue-number -->

cooperlees committed:
bpo-38483 Add support for ~/.venvrc

- Add argparse.ArgumentDefaultsHelpFormatter so user can see overloads
- Add new CliDefaults NamedTuple with defaults
- Check for ~/.venvrc + parse if exists
- Check args, if true or false bool otherwise string
- Overload by create NamedTuple passing kwargs
- Update docs

Test:
- Run `./python.exe -m venv --help`
- Add unit tests fir new function

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

cooperlees committed:
Move to using suggested venv.ini

vsajip commented on Doc/using/venv-create.inc:
```diff
@@ -132,3 +132,24 @@ detail (typically a script or shell function will be used).
+
+The venv module supports an ini config file in the following locations:
+
+* ``$HOME/.venv.ini`` on Linux/Unix (+ Darwin/Mac OSX) platforms
```
This location should be `$HOME/.config/python/venv.ini` to match the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html).

merwok replied:
`$HOME/.config` is the default value for `$XDG_CONFIG_HOME`; if a Python tool follows the spec it should follow it fully IMO.

Could the file be named `venv.cfg` rather than `.ini`?  Windows ini format is an ill-defined format with many variations, whereas `cfg` is more evocative of `configparser` module.

vsajip commented on Doc/using/venv-create.inc:
```diff
@@ -132,3 +132,24 @@ detail (typically a script or shell function will be used).
+The venv module supports an ini config file in the following locations:
+
+* ``$HOME/.venv.ini`` on Linux/Unix (+ Darwin/Mac OSX) platforms
+* ``%APPDATA%\Python\venv\venv.ini'`` on Windows platforms
```
Perhaps `%APPDATA%\Python\venv.ini` - there should be no need for an extra `venv` subdirectory for just one file.

vsajip commented on Doc/using/venv-create.inc:
```diff
@@ -132,3 +132,24 @@ detail (typically a script or shell function will be used).
+* ``$HOME/.venv.ini`` on Linux/Unix (+ Darwin/Mac OSX) platforms
+* ``%APPDATA%\Python\venv\venv.ini'`` on Windows platforms
+
+to overload CLI defaults. This allows any CLI argument mentioned above
```
This usage of 'overload' might be confusing - perhaps 'supply' might be used instead. Anything provided on the command line should override a corresponding setting in `venv.ini`. The values in `venv.ini` only override those that would be otherwise set in the code in the absence of appropriate CLI arguments. So, I suggest you update the comment to make it clearer that:

* If arguments are provided on the command line, they will be used to set parameter values.
* Otherwise, if a setting is specified in `venv.ini`, it will be used.
* In the absence of both of the above, a suitable default will be applied in the code.

vsajip commented on Doc/using/venv-create.inc:
```diff
@@ -132,3 +132,24 @@ detail (typically a script or shell function will be used).
+    [venv]
+    upgrade_deps = true
+
+This would result in ``--upgrade-deps`` to defualt to True.
```
defualt -> default

vsajip commented on Doc/using/venv-create.inc:
```diff
@@ -132,3 +132,24 @@ detail (typically a script or shell function will be used).
+Please use ``--help`` to test ``venv.ini`` applied defaults
+
+.. versionadded:: 3.9
+   Add support for venv.ini to overload CLI argument defaults
```
overload -> supply

vsajip commented on Lib/test/test_venv.py:
```diff
@@ -380,6 +381,34 @@ def test_deactivate_with_strict_bash_opts(self):
+    """ Test various CLI related functions """
+
+    BAD_VENV_INI = "|not_venv_or_ini|\nkey = value\n"
+    GOOD_VENV_INI = "[venv]\nupgrade_deps = TrUe"  # case on purpose for test
```
The section should be `DEFAULT` rather than `venv`, as it's the only one.

vsajip commented on Lib/venv/__init__.py:
```diff
@@ -4,19 +4,34 @@
 CORE_VENV_DEPS = ('pip', 'setuptools')
+VENV_INI_DEFAULT = '~/.venv.ini'
+if sys.platform == "win32":
+    VENV_INI_DEFAULT = os.path.expandvars(r'%APPDATA%\Python\venv\venv.ini')
```
Would redo this section as
```
if os.name == 'nt':
    _dir = os.path.expanduser('~/.config')
else:
    _dir = os.path.expandvars(r'%APPDATA%\Python')
os.makedirs(_dir, exist_ok=True)
VENV_INI_DEFAULT_PATH = os.path.join(_dir, 'venv.ini')
del _dir
```

vsajip commented on Lib/venv/__init__.py:
```diff
@@ -409,8 +424,34 @@ def create(env_dir, system_site_packages=False, clear=False,
     builder.create(env_dir)
 
+
+def get_default_args(venv_ini_default=VENV_INI_DEFAULT):
```
venv_ini_default -> venv_ini_path
VENV_INI_DEFAULT -> VENV_INI_DEFAULT_PATH.

Might be an idea to just default to `None` here, and in the body do
```
if venv_ini_path is None:
    venv_ini_path = DEFAULT_VENV_INI_PATH
```
I generally do this whenever the default value is not a simple constant (it looks on the surface like a constant, but involves some platform-dependent computation to derive).

vsajip commented on Lib/venv/__init__.py:
```diff
@@ -409,8 +424,34 @@ def create(env_dir, system_site_packages=False, clear=False,
+def get_default_args(venv_ini_default=VENV_INI_DEFAULT):
+    """ Check if a venv.ini exists in the platform location
+        + override defaults if it exists. Use `--help` to test ini config """
+    venvini_path = os.path.expanduser(venv_ini_default)
```
No, compute it completely above, as I've suggested in an earlier comment.

vsajip commented on Lib/venv/__init__.py:
```diff
@@ -4,19 +4,34 @@
 logger = logging.getLogger(__name__)
 
 
+class CliDefaults(NamedTuple):
```
Suggest renaming CliDefaults -> Defaults. They're just defaults to use for the underlying API, and may or may not come from a command line.

vsajip commented on Lib/venv/__init__.py:
```diff
@@ -409,8 +424,34 @@ def create(env_dir, system_site_packages=False, clear=False,
+
+    cp = configparser.ConfigParser()
+    cp.read(venvini_path)
+    if not cp.has_section("venv"):
```
venv -> DEFAULT

vsajip commented on Lib/venv/__init__.py:
```diff
@@ -409,8 +424,34 @@ def create(env_dir, system_site_packages=False, clear=False,
+        else:
+            arg_overloads[name] = value
+
+    return CliDefaults(**arg_overloads)
```
This logic will fail with a `TypeError` if the ini contains a variable which is not a valid tuple field name, or a typo in the value. As this is user-specified, untrusted input, checking needs to be bullet-proof.

Would redo as a loop over the known/expected field names, with checking for expected types, doing the 'true'/ True etc. conversions, checking for duplicates (raise `ValueError`) and so on.

vsajip commented on Misc/NEWS.d/next/Library/2019-10-15-13-03-59.bpo-34556.K2XyWH.rst:
```diff
@@ -0,0 +1 @@
+Add ``venv.ini`` support to venv module to overload CLI argument defaults. Patch by Cooper Ry Lees
```
overload -> supply

vsajip commented on Lib/test/test_venv.py:
```diff
@@ -380,6 +381,34 @@ def test_deactivate_with_strict_bash_opts(self):
+        expected = venv.CliDefaults(upgrade_deps=True)
+        with open(venv_ini_path, "w") as vfp:
+            vfp.write(self.GOOD_VENV_INI)
+        self.assertEqual(venv.get_default_args(venv_ini_path), expected)
```
I would add more failure cases than just a bad section. Cases to include (both successes and failures):

* Typos in bool values (e.g. `upgrade_deps = flse` - `TypeError` should be raised)
* Duplicated values (`ValueError` should be raised)
* Spurious key values (ignored)
* Multiple values set at once

cooperlees committed:
I did all that I could from your review @vsajip.

I do want to state, I think running os.mkdirs on import is bad form. What do you think about moving the dir creation into get_default_args or somewhere else non global?

Changes:
- Raise value error for bool options with typos / non true/false values
  - I could have dynamically look for bool objects via `_field_types`, but it is a private OrderedDict so decided against. I could use it if you want via set comprehension for the bool_keys list
- ConfigParser looks for duplicates keys as strict mode is on by default
  - Added test for that
- I couldn't get the DEFAULT section to work
  - Everywhere I looked everyone uses a different section with DEFAULT section to provide default values.
  - I couldn't get it to work how you've requested - Do you have some example code anywhere using it the way you'd like me to use it here?
- I didn't understand what "Multiple values set at once" means ... I feel like configparser covers this
- Add tests to cater for your needs - I feel configparser exceptions are better than a generic ValueError where I have used them
- Corrected documentation

ambv commented on Lib/venv/__init__.py:
```diff
@@ -409,8 +429,52 @@ def create(env_dir, system_site_packages=False, clear=False,
+        try:
+            getattr(defaults, name)
+        except AttributeError:
+            continue
```
At this point `if not hasattr(defaults, name): continue` is clearer. Better yet, to ensure you're only allowing overriding things that are proper fields on the Defaults namedtuple, do:
```py3
if name not in Defaults._fields:
    continue
```

ambv commented on Lib/venv/__init__.py:
```diff
@@ -409,8 +429,52 @@ def create(env_dir, system_site_packages=False, clear=False,
+                raise ValueError(
+                    f"{name} is a bool. Config can only be set to "
+                    + f"False or True only. Not '{value}'"
+                )
```
Instead of this entire custom logic, you can just say:
```py3
read_kwargs[name] = cp["venv"].getboolean(key)
```
It will also raise ValueError on invalid values.

cooperlees replied:
Much cleaner. Thanks.

ambv commented on Lib/venv/__init__.py:
```diff
@@ -409,8 +429,52 @@ def create(env_dir, system_site_packages=False, clear=False,
+    bool_keys = {
+        "clear", "copies", "system_site_packages", "upgrade_deps",
+        "use_symlinks", "without_pip"
+    }
```
To make sure this stays current:
```py3
bool_keys = {
    key
    for key, value_type in Defaults._field_types.items()
    if value_type is bool
}
```

cooperlees replied:
I did discuss above this option above. Cool, if you think it's not crazy I'll do it too.

ambv commented on Lib/venv/__init__.py:
```diff
@@ -4,18 +4,38 @@
 logger = logging.getLogger(__name__)
 
+if os.name == "nt":
+    _dir = os.path.expandvars(r'%APPDATA%\Python')
```
Pro-tip: if your raw-string is not a regular expression and you write R'...' instead of r'...' then GitHub will highlight it better.

cooperlees replied:
TIL

cooperlees replied:
It seems in `vscode` too in case you didn't know :D

ambv reviewed (commented):
This looks good to me overall. My only concerns are:

1. Are the configuration paths okay? For example on macOS `appdirs.user_config_dir('python')` will give you `~/Library/Application Support/python` and not `~/.config/python`.
2. Should there be a way to specify a custom `--config=PATH`?

cooperlees committed:
- Move mkdirs into get_default_args and not during global
- Apply @ambv suggestions with better NamedTuple + configparser API usage

cooperlees committed:
Change ini file + tests to use DEFAULT and show failures
